app/src/main/java/com/demoapp/database/DataRepository.kt

```diff
@@ -1,64 +1,54 @@
 package com.demoapp.database
 
-import android.util.Log
+
 import com.demoapp.model.CategoryEntity
 import com.demoapp.model.DataEntity
 import com.demoapp.model.DataResponse
 import com.demoapp.retrofit.RetrofitInterface
-import retrofit2.Call
-import retrofit2.Callback
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
 
-class DataRepository(val dataDuo: DataDuo, val mService: RetrofitInterface) {
+class DataRepository @Inject constructor(
+    private val dataDuo: DataDuo,
+    private val mService: RetrofitInterface
+) {
 
-    fun getLocalData(result : (List<DataEntity>) -> Unit) {
-        var resultList = arrayListOf<DataEntity>()
+    suspend fun getLocalData(): List<DataEntity> {
         val categories = dataDuo.getAllCategories()
         val dataList = dataDuo.getData()
 
-        categories.forEach { category ->
-            resultList.add(
-                DataEntity(
-                    name = "",
-                    description = "",
-                    price = 0.0,
-                    image_link = "",
-                    product_id = "",
-                    category_name = category.name,
-                    is_category = true
-                )
-            )
-
-            resultList.addAll(dataList.filter {
-                it.category_name.equals(
-                    category.name,
-                    ignoreCase = true
+        return arrayListOf<DataEntity>().apply {
+            categories.forEach {
```

Comment:
No foreach allowed for data transformation

Reply:
Flatmap used here instead of foreach and arraylist